README.md
# AppFlow

**Read the documentation on [appflow.rtfd.io](https://appflow.readthedocs.io)!**

**For the playbooks, visit the [appflow-playbooks repository](https://github.com/ttssdev/appflow-playbooks)**

Get an invite and join the conversations!  
[![Slack Status](https://static1.squarespace.com/static/53f68e19e4b0f401658fbb93/58b99eee725e2580fa698860/58b9a61603596ea54d1c5035/1488819693257/slack-logo-01.png?format=100w)](https://appflow-community.ttss.ch)  

AppFlow is a multitenant environment automation tool based on Ansible.

AppFlow is your transparent platform that incorporates the various digital infrastructures into a continuous workflow. At the same time, it creates a consistent environment for the entire development process, saving time and money. 
AppFlow is an open source developer and DevOps application that maps the entire development process (continuous deployment & provisioning) from development to test & staging to production. This ensures that the systems being managed on the different infrastructures are deployed in the same way. At the same time, AppFlow creates a consistent environment (on-premise, in the cloud or on bare metal) throughout the development process. The administrators and developers (DevOps) can therefore independently control the entire toolchain. The automation process is created using Ansible playbooks and contains the admin and developer code, which is stored in a git repository (encrypted). The code (AppFlow) runs on any UNIX-like operating system where Python can run, including Linux, BSD, and OSX. 
Orchestration of the AppFlow code requires only SSH access to the respective systems. 
AppFlow has been designed and optimized from DevOps for DevOps.  

## Features

Provisioning:

* Multitenant architecture (different teams with different environments)
* Supports `development`, `testing`, `staging` and `production`
* All configuration files are encrypted in git with `ansible-vault`
* Provision all nodes with one command

Development:

* Provides a Vagrant based development environment called `atlantis`
* Code locally on any Unix-like system or Windows (cygwin)

Deployment:

* Made for [Bedrock](https://roots.io/bedrock/) projects and [bedrock-capistrano](https://github.com/roots/bedrock-capistrano) deployments
* Deploy and rollback with one command

Infrastructure:

* Builtin [Percona XtraDB Cluster 5.6](https://www.percona.com/software/mysql-database/percona-xtradb-cluster) and [GlusterFS](http://www.gluster.org) support for sharing web uploads on multiple nodes
* Easy development environments with [Vagrant](http://www.vagrantup.com/)
* Easy server provisioning with [Ansible](http://www.ansible.com/) (Ubuntu 16.04, PHP 5.6/7)

## Technologies
The technology behind AppFlow uses countless best-in-class programs and maps them in a toolbox. The software accesses an extensive repository of various freeware packages. Out-of-the-box, many enterprise features already exist, such as:


 - Load balancing
 - Apache / PHP
 - Web Accelerator / Caching / PageSpeed
 - Distributed file system
 - Master-Master Database
 - Backup & Monitoring Integration
 - Jailkit - chroot
 - ...

### Installation

Appflow is hosted on PiP using python3.
`pip3 install appflow` 
will install appflow.
To start using it you first need to *initialize* it:
`appflow init`
follow the onscreen instructions to set it up!

### Developers

#### Contribute a new feature

 - Create a new issue, e.g. #XX new superfeature
 - Create local branch: `git checkout -b XX-new-superfeature`
 - Code on it.
 - Push it to remote as new branch: `git push -u origin XX-new-superfeature`
 - Create new pull request (`base: master ... compare: XX-new-superfeature`)

#### Get all vars

`ansible all -m setup --tree /tmp/facts -i examples/YOUR_TENANT/local/inventory -a "filter=ansible_distribution*"`

## Contributing

Contributions are welcome from everyone. 

Join us! [![Slack Status](https://static1.squarespace.com/static/53f68e19e4b0f401658fbb93/58b99eee725e2580fa698860/58b9a61603596ea54d1c5035/1488819693257/slack-logo-01.png?format=100w)](https://appflow-community.ttss.ch) 


docs/changes.rst
Changes
=======

Appflow 1.0.1.5
~~~~~~~~~~~~~~~
Minor update introducing:
    - Appflow-playbooks versioning
    - Appflow-playbooks selectable branch
    - `appflow version` will now yield also the playbooks version:
        ``Appflow Version: 1.0.1.5``
        ``Playbooks Version 1.0.0``

Appflow 1.0.1.4
~~~~~~~~~~~~~~~

Released a little update including
    - introduced appflow version command
    - introduced appflow provision debug
    - fixed newline bug in appflow hosts

to update just use Pip: `pip3 install -U appflow`


Screencasts
~~~~~~~~~~~

To complement the documentation, we've added screencasts recorded on `asciinema.org <http://asciinema.org/>`__

Screencasts include the walkthrought of the basic setups and features of Appflow.
They include:
    
    - `Installation`_
    - `Basic setup`_
    - `Atlantis setup`_
    - `Atlantis provision`_
    - `Project provision`_


Varnish Grace mode
~~~~~~~~~~~~~~~~~~

`Grace mode <https://varnish-cache.org/docs/5.1/users-guide/vcl-grace.html>`_ has been a long expected feature and went production ready with
recent appflow-playbooks commit `f0d7f3817ffb1b2354f0c24a98e3dac37b72202d <https://github.com/ttssdev/appflow-playbooks/commit/f0d7f3817ffb1b2354f0c24a98e3dac37b72202d>`_.

This special operation mode in Varnish allows a website to remain online and running also when the backend components, like
MySQL or Apache2, are not running. This works beccause Varnish will serve all the web content directly from it's cache.

Backend operations like login to CMS or similar will obviously not work as expected but at least the public facing part of the
website will not be offline, so no 404s or similar for end-users.

Grace period is set by default to 6h but can be extended to one week or whatever your preference is, the main advantage
of Grace is that your site remains up when bad things happen and you'll get a time buffer for fixing whatever issue happened
to the backend.

Grace mode will be enabled by default, for any environment, if you perform:

::

    % appflow update
    % appflow provision --tags varnish-conf,apache2-conf,mysql

It's important to note that Grace mode is active by default from Varnish 5.1 upwards so if you're on 3.x
you'll need to upgrade varnish first, this can be done, in development, by:

::

    % appflow update
    % ssh atlantis "sudo service varnish stop"
    % appflow provision --tags varnish,apache2-conf,mysql

You also need to set the Varnish version in ``group_vars/webservers``:

::

    #
    # Varnish
    #
    conf_varnish_version: 51
    ...

a complete config setting for Varnish 5.1 and Grace would look like:

::

    #
    # Varnish
    #
    conf_varnish_version: 51
    conf_varnish_listen_port: 6081
    conf_varnish_listen_admin_port: 60821
    conf_varnish_backend_default: |
    .host = "127.0.0.1";
    .port = "8080";
    .max_connections = 800;
    conf_varnish_acl_purge: |
    "127.0.0.1";
    "localhost";
    "192.168.80.2";
    conf_varnish_vcl_recv: |
    {% if conf_lbtier_enable == false %}
    remove req.http.X-Forwarded-For;
    set req.http.X-Forwarded-For = client.ip;
    {% endif %}

    if (req.url ~ "/wp(-login|-admin|-cron|login|-comments-post.php)" ) {
        return (pass);
    }

    if (req.http.Cache-Control ~ "no-cache") {
        return (pass);
    }

    # Remove client-side cookies.
    set req.http.Cookie = regsuball(req.http.Cookie, "(^|;\s*)(_[_a-z]+|has_js|utmctr|utmcmd.|utmccn.|WT_FPC|_hjIncludedInSample)=[^;]*", "");

    # Remove a ";" prefix, if present.
    set req.http.Cookie = regsub(req.http.Cookie, "^;\s*", "");

    # Are there cookies left with only spaces or that are empty?
    if (req.http.cookie ~ "^\s*$") {
        unset req.http.cookie;
    }

    conf_varnish_vcl_backend_response: |
    # Allow stale content, in case the backend goes down.
    # make Varnish keep all objects for 6 hours beyond their TTL
    set beresp.grace = 6h;
    # set beresp.grace = 2m;

    conf_varnish_vcl_fetch: |
    # set beresp.grace = 2m;

    # If the URL is for one of static images or documents, we always want them to be cached.
    if (beresp.status == 200 && req.url ~ "\.(ico|jpe?g|jpe|gif|png|webp|svg|css|js)$") {
    # Cookies already removed.
    # Cache the page for 10 days.
        set beresp.ttl = 10d;
    # Remove existing Cache-Control headers.
        remove beresp.http.Cache-Control;
    # Set new Cache-Control headers for browser to store cache for 7 days.
        set beresp.http.Cache-Control = "public, max-age=604800";
    }

    # Cache 404 responses for 15 seconds.
    if (beresp.status == 404) {
        set beresp.ttl = 15s;
        set beresp.grace = 15s;
    }

    conf_varnish_vcl_deliver: |
    # For security and asthetic reasons, remove some HTTP headers before final delivery.
    unset resp.http.Server;
    unset resp.http.X-Powered-By;
    unset resp.http.Via;
    unset resp.http.X-Varnish;
    Once Varnish has been updated and Grace mode has been enabled you could test if it's working correctly:

    
Open one of the web projects you're hosting on Atlantis in the browser, everything should be there.

::

    % curl http://atlantis:8080/health.php
    MySQL running

    % ssh atlantis "sudo varnishadm backend.list"
    Backend name                   Admin      Probe
    boot.default                   probe      Healthy

    % ssh atlantis "sudo service mysql stop"

    % ssh atlantis "sudo varnishadm backend.list"
    Backend name                   Admin      Probe
    boot.default                   probe      Sick

Open the previous web project again in the browser, the website should be online as usual via Grace mode.

The same concept applies also to production where we have three or more nodes.


.. start-badges

.. _Installation: https://asciinema.org/a/0lglEIPiYhsceMExzOKHBUcdZ?autoplay=1&speed=1
.. _Basic setup: https://asciinema.org/a/VRlp5YqiT4gvKXrYFYZW9Oz3l?autoplay=1&speed=1

rest-server.py
#!/usr/bin/python3

from flask import Flask
from flask import make_response
from flask import request

import appflow.AppflowAnsible as apansible
import appflow.AppflowTools as tools
import appflow.AppflowYaml as apyaml

app = Flask(__name__)

# GET -> get
# POST -> all provision commands
# PUT   -> add
# PATCH -> set
# DELETE -> rm

# 204 NO CONTENT – [DELETE/POST/PUT/PATCH]
# 400 INVALID REQUEST – [DELETE/POST/PUT/PATCH]
# 400 MISSING ARGUMENTS [maybe for the registration/signin]
# 401 UNAUTHORIZED -> Maybe if we use api tokens???
# 403 FORBIDDEN -> Maybe if we use api tokens???
# 404 NOT FOUND -> Invalid Command or syntax
# 405 METHOD NOT ALLOWED -> Invalid Command or syntax
# 500 INTERNAL SERVER ERROR

# use db;
# select * from users \G;
# CREATE TABLE users(username TEXT, mail TEXT, password TEXT, salt TEXT, tenant TEXT, api_key TEXT, api_quota INT, user_enc TEXT, mail_enc TEXT, api_key_enc TEXT);

# https://blog.miguelgrinberg.com/post/restful-authentication-with-flask
# https://github.com/miguelgrinberg/REST-auth
# https://blog.miguelgrinberg.com/post/designing-a-restful-api-with-python-and-flask --> Securing a RESTful web service
# https://www.owasp.org/index.php/REST_Security_Cheat_Sheet#Access_Control


@app.route("/appflow/get", methods=['GET'])
def get():
    try:
        file_name = request.args.get('file')
        key = request.args.get('key')

        response = apyaml.get_value(file_name, key)
        if 'Error' in response:
            if 'Invalid' in response:
                status_code = 400
            elif 'Bad Syntax' in response:
                status_code = 404
            elif 'Not Found' in response:
                status_code = 204
            else:
                status_code = 500
        else:
            status_code = 200
        return make_response(response, status_code)
    except Exception as exception:
        return make_response(str(exception), 500)


@app.route("/appflow/set", methods=['PATCH'])
def set():
    try:
        file_name = request.args.get('file')
        key = request.args.get('key')
        value = request.args.get('value')
        response = apyaml.set_value(file_name, key, value)
        if 'Error' in response:
            if 'Invalid' in response:
                status_code = 400
            elif 'Bad Syntax' in response:
                status_code = 404
            elif 'Not Found' in response:
                status_code = 204
            else:
                status_code = 500
        else:
            status_code = 200
        return make_response(response, status_code)

    except Exception as exception:
        return make_response(str(exception), 500)


@app.route("/appflow/add",  methods=['PUT'])
def add():
    try:
        file_name = request.args.get('file')
        key = request.args.get('key')
        value = request.args.get('value')
        response = apyaml.add_value(file_name, key, value)
        if 'Error' in response:
            if 'Invalid' in response:
                status_code = 400
            elif 'Bad Syntax' in response:
                status_code = 404
            elif 'Not Found' in response:
                status_code = 204
            else:
                status_code = 500
        else:
            status_code = 200
        return make_response(response, status_code)

    except Exception as exception:
        return make_response(str(exception), 500)


@app.route("/appflow/rm",  methods=['DELETE'])
def rm():
    try:
        file_name = request.args.get('file')
        key = request.args.get('key')
        response = apyaml.rm_value(file_name, key)
        if 'Error' in response:
            if 'Invalid' in response:
                status_code = 400
            elif 'Bad Syntax' in response:
                status_code = 404
            elif 'Not Found' in response:
                status_code = 204
            else:
                status_code = 500
        else:
            status_code = 200
        return make_response(response, status_code)

    except Exception as exception:
        return make_response(str(exception), 500)


@app.route("/appflow",  methods=['POST'])
def command():
    try:
        command = request.args.get('command', default='', type=str)
        tenant = request.args.get('tenant', default='', type=str)
        env = request.args.get('env', default='', type=str)
        tags = 'tags=' + request.args.get('tags', default='', type=str)
        skip_tags = 'skip_tags=' + \
            request.args.get('skip_tags', default='', type=str)
        limit = 'limit=' + request.args.get('limit', default='', type=str)
        ask_sudo_pass = 'ask_sudo_pass=' + \
            request.args.get('ask_sudo_pass', default='', type=str)
        user = 'user=' + request.args.get('user', default='', type=str)

        return {
            'reset': tools.git_reset(tenant, env),
            'status': tools.git_status(tenant, env),
            'checkin': tools.git_check_in(tenant, env),
            'checkout': tools.git_check_out(tenant, env),
            'encrypt': apansible.encrypt(tenant, env),
            'decrypt': apansible.decrypt(tenant, env),
            'tags': apansible.tags(tenant, env),
            'provision': apansible.provision(tenant, env, tags, skip_tags, limit, ask_sudo_pass, user)
        }.get(env, 'development')
    except Exception as exception:
        return make_response(str(exception), 500)


# @app.route("/appflow/signin",  methods=['POST'])
# @app.route("/appflow/signup/get/get_user_data",  methods=['POST'])
# @app.route("/appflow/signin/get/recover_mail",  methods=['POST'])
# @app.route("/appflow/signin/get/recover_username",  methods=['POST'])
# @app.route("/appflow/signin/get/recover_api_key",  methods=['POST'])
# @app.route("/appflow/signin/change_pass",  methods=['POST'])
# @app.route("/appflow/signin/change_mail",  methods=['POST'])
if __name__ == "__main__":
    app.run(host='0.0.0.0', port=8000)
